source/server/drain_manager_impl.h

```diff
@@ -29,30 +29,40 @@ class DrainManagerImpl : Logger::Loggable<Logger::Id::main>, public DrainManager
                    Event::Dispatcher& dispatcher);
 
   // Network::DrainDecision
-  bool drainClose() const override;
-  Common::CallbackHandlePtr addOnDrainCloseCb(DrainCloseCb cb) const override;
+  bool drainClose(Network::DrainDirection scope) const override;
+  Common::CallbackHandlePtr addOnDrainCloseCb(Network::DrainDirection direction,
+                                              DrainCloseCb cb) const override;
 
   // Server::DrainManager
-  void startDrainSequence(std::function<void()> drain_complete_cb) override;
-  bool draining() const override { return draining_; }
+  void startDrainSequence(Network::DrainDirection direction,
+                          std::function<void()> drain_complete_cb) override;
+  bool draining(Network::DrainDirection direction) const override {
+    return draining_.load().first && direction <= draining_.load().second;
+  }
   void startParentShutdownSequence() override;
   DrainManagerPtr
   createChildManager(Event::Dispatcher& dispatcher,
                      envoy::config::listener::v3::Listener::DrainType drain_type) override;
   DrainManagerPtr createChildManager(Event::Dispatcher& dispatcher) override;
 
 private:
-  void addDrainCompleteCallback(std::function<void()> cb);
+  void addDrainCompleteCallback(Network::DrainDirection direction, std::function<void()> cb);
 
   Instance& server_;
   Event::Dispatcher& dispatcher_;
   const envoy::config::listener::v3::Listener::DrainType drain_type_;
-
-  std::atomic<bool> draining_{false};
-  Event::TimerPtr drain_tick_timer_;
-  MonotonicTime drain_deadline_;
+  using DrainPair = struct {
+    bool first;
+    Network::DrainDirection second;
+  };
+  std::atomic<DrainPair> draining_{DrainPair{false, Network::DrainDirection::None}};
```

Comment:
At this point I would question the choice of atomic and start using the mutex. Specifically it is not read in an atomic fashion any more, as you consistently do two loads to examine each member. Two loads can have an interleaving write.

Reply:
Ah, good catch. I meant to go back and assign the load to an intermediate variable and then evaluate but I lost track